### PR TITLE
pathlib: Patch pathlib to work around some bugs (fixes #8263 #7295)

### LINKED ISSUE #8263: Codepage issue on Windows 8.1
**Describe the bug**
I receive a build failure when building `harfbuzz` through `vcpkg`. `vcpkg` maintainers believe it is a `meson` upstream bug with codepages. The downstream bug is microsoft/vcpkg#15916. I have no ability to file a sane bug report, since I know nothing. However, I have logs.

**To Reproduce**
I can only reproduce this when building harfbuzz with vcpkg, using `./vcpkg install harfbuzz`. I don't receive this error when building harfbuzz with the project's native meson instructions without vcpkg. These builds are on the same system, but my native build uses mingw-w64's gcc, while the vcpkg build uses Visual Studio. The compiler isn't yet invoked when the build error occurs, so I don't believe this difference of compilers is the cause.

**Expected behavior**
I expected it to build cleanly. This is the log:

```
[3/66] "C:\Users\adm\Downloads\vcpkg\downloads\tools\python\python-3.9.0-x64\python.exe" "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\meson.py" "--internal" "depscan" "src/harfbuzz.dll.p/harfbuzz.dat" src/harfbuzz.dll.p/depscan.dd ../src/f4d3aa1127-a820a32874.clean/src/hb-aat-layout.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-aat-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-blob.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-buffer-serialize.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-buffer.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-common.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-draw.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-face.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-fallback-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-font.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-number.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-cff1-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-cff2-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-color.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-face.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-font.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-layout.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-math.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-meta.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-metrics.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-name.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-arabic.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-default.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-hangul.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-hebrew.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-indic-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-indic.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-khmer.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-myanmar.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-thai.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-use-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-use.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-vowel-constraints.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-fallback.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-normalize.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-tag.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-var.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-set.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shape-plan.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shaper.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-static.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-style.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ucd.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-unicode.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ft.cc
FAILED: src/harfbuzz.dll.p/depscan.dd 
"C:\Users\adm\Downloads\vcpkg\downloads\tools\python\python-3.9.0-x64\python.exe" "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\meson.py" "--internal" "depscan" "src/harfbuzz.dll.p/harfbuzz.dat" src/harfbuzz.dll.p/depscan.dd ../src/f4d3aa1127-a820a32874.clean/src/hb-aat-layout.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-aat-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-blob.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-buffer-serialize.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-buffer.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-common.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-draw.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-face.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-fallback-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-font.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-number.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-cff1-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-cff2-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-color.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-face.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-font.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-layout.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-map.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-math.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-meta.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-metrics.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-name.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-arabic.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-default.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-hangul.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-hebrew.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-indic-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-indic.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-khmer.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-myanmar.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-thai.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-use-table.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-use.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-complex-vowel-constraints.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-fallback.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape-normalize.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-tag.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ot-var.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-set.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shape-plan.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shape.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-shaper.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-static.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-style.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ucd.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-unicode.cc ../src/f4d3aa1127-a820a32874.clean/src/hb-ft.cc
Traceback (most recent call last):
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\meson.py", line 29, in <module>
    sys.exit(mesonmain.main())
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\mesonmain.py", line 229, in main
    return run(sys.argv[1:], launcher)
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\mesonmain.py", line 218, in run
    return run_script_command(args[1], args[2:])
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\mesonmain.py", line 166, in run_script_command
    return module.run(script_args)
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\scripts\depscan.py", line 91, in run
    return scanner.scan()
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\scripts\depscan.py", line 64, in scan
    self.scan_file(s)
  File "C:\Users\adm\Downloads\vcpkg\downloads\tools\meson\meson-f5871f434a5b768ad9fcafa797a6db0286421842\mesonbuild\scripts\depscan.py", line 37, in scan_file
    for line in pathlib.Path(fname).read_text().split('\n'):
  File "pathlib.py", line 1256, in read_text
  File "encodings\cp1252.py", line 23, in decode
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 36961: character maps to <undefined>
```

Attached: [src\harfbuzz.dll.p\harfbuzz.dat](https://github.com/mesonbuild/meson/files/5883593/harfbuzz.txt)
Attached: [package-x64-windows-rel-out.log](https://github.com/mesonbuild/meson/files/5883605/package-x64-windows-rel-out.log), probably no need to look at this

**system parameters**
* native build
* Windows 8.1
* Python 3.9.1
* meson 0.56.2
* ninja 1.10.2

### PR DESCRIPTION
**Pull request by mensinda**

### CODE REVIEW

**Comment by eli-schwartz** [comment]
Test failures seem very relevant:
```
_pickle.PicklingError: Can't pickle <class 'pathlib.PosixPath'>: attribute lookup PosixPath on pathlib failed
```

I don't understand the PR anyway, per @jpakkane 
> Rather than monkeypatching, it would be better to change the call sites. The thing about monkeypatching is that it always comes back to bite you in the ass.

But here we are, monkeypatching `sys.modules['pathlib']`.

Last time we were hacking around pathlib, we did this everywhere:

```
-from pathlib import Path
+from mesonbuild._pathlib import Path
```

**Comment by eli-schwartz** [comment]
Given this (#8263) is a "feature" of python, not a bug, it feels wrong to modify pathlib as we'd need to keep this code around "literally forever". Why is passing explicit encoding= so unpalatable to you?

**Comment by jpakkane** [comment]
> we have no good way of ensuring that we actually always pass encoding='utf-8'.

One can write a unit that that regexes all `.py` files for `read_text` that is not followed by `encoding`. It's not 100% reliable but for this 95% should be good enough.

**Comment by mensinda** [comment]
> One can write a unit that that regexes all .py files for read_text that is not followed by encoding. It's not 100% reliable but for this 95% should be good enough.

Yeah, this might be the better solution. Let's see if I can get something to work.

**Comment by mensinda** [comment]
OK, changed **all** the `open()` and `{read,write}_text()` to always use `encoding=` (and have a script check it).

**Comment by jpakkane** [comment]
The `_pathlib` file should be removed then.

**Comment by eli-schwartz** [comment]
> NEW  The module 'mesonbuild._pathlib' imports itself.

Erm, what. This check looks quite broken, lol, _pathlib is not pathlib

> NEW   Import of 'T' is not used.

Accurate analysis...

Given the overridden _pathlib.py is now only here to fix these unusual ImDisk issues, we might want to defer it to upstream python?

**Comment by mensinda** [comment]
@jpakkane The `_pathlib` module is still required to work around the actual python bug with `.resolve()`.

https://bugs.python.org/issue31842
https://github.com/mesonbuild/meson/issues/7295

**Comment by mensinda** [comment]
Fixed, the unused typing issue.

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -462,10 +462,8 @@ def detect_vs_dep_prefix(self, tempfilename):
             return open(tempfilename, 'a', encoding='utf-8')
         filename = os.path.join(self.environment.get_scratch_dir(),
                                 'incdetect.c')
-        with open(filename, 'w') as f:
-            f.write('''#include<stdio.h>
-int dummy;
-''')
+        with open(filename, 'w', encoding='utf-8') as f:
+            f.write('#include<stdio.h>\nint dummy;\n')
```
Comment (eli-schwartz) [inline mesonbuild/backend/ninjabackend.py:466]: Unrelated string change.

Also I think the change looks ugly!

File: run_custom_lint.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+def check_missing_encoding(lines: T.List[str], path: str) -> int:
+    errors = 0
+    functions = ['read_text', 'write_text', 'open']
+    for num, line in enumerate(lines):
+        for func in functions:
+            l = line
+
+            # Skip ignored lines
+            if '[ignore encoding]' in l:
+                continue
```
Comment (eli-schwartz) [inline run_custom_lint.py:20]: Why do we need magic comments specific to this script, in order to flag allowed uses of encoding='something other than utf-8'?

If something intentionally other than utf-8 is being used, then it's obviously intentional and therefore obviously fine (at least, ***it will behave the same on Linux and Windows***). So we should just permit it. No need to later on check that the encoding is specifically either 'ascii' or 'utf-8'.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/backend/ninjabackend.py (hunk 1 above)
Comment (dcbaker) [reply mesonbuild/backend/ninjabackend.py:466]: textwrap.dedent is best for this I think

File: run_custom_lint.py (hunk 2 above)
Comment (mensinda) [reply run_custom_lint.py:20]: Fine, I removed the special cases for `'ascii'` and `'utf-8'`, but this is still needed for stuff like `tarfile.open(`.

File: mesonbuild/backend/ninjabackend.py (hunk 1 above)
Comment (mensinda) [reply mesonbuild/backend/ninjabackend.py]: switched to `dedent`.

File: mesonbuild/_pathlib.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+import os
+
```
Comment (jpakkane) [inline mesonbuild/_pathlib.py:17]: It does not seem possible to do this without monkey patching, so that's what we need to do then. However the implementation needs some fixes:

- the name should not be `_pathlib` as leading underscores in names are confusing and can lead to problems (as can be seen with the CI), instead it should be something more dangerous sounding like `pybugXXXXworkaround`
- The file should have a comment at the top explaining what it is, why it's needed, have links to the relevant bug reports, a notice that no new code or functionality should be added to it and also when the workaround can be removed.

Comment (mensinda) [reply mesonbuild/_pathlib.py]: Done.

Comment (mensinda) [reply mesonbuild/_pathlib.py]: Well, I guess renaming the file did not fix the `1 for Module imports itself` warning...

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: That seems like an LGTM bug then. Let's just rename it `pathlibworkaround` and call it a day.

Comment (mensinda) [reply mesonbuild/_pathlib.py]: So, do you want me to rename it again, or does the current name work?

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: According to the web ui the module name is `zzz_pathlibworkaround` which is not a great name ...

Comment (mensinda) [reply mesonbuild/_pathlib.py]: But, then it looks like a regular meson module and blends in with the rest of them, which goes against the point of making it look scary. Also, we already have the `_typing` module...

Would you be ok with renaming it to `zzzpathlibworkaround`? Also which web ui is complaining about the name?

Comment (eli-schwartz) [reply mesonbuild/_pathlib.py]: I believe LGTM is complaining that whatever the name of your module, it's being inserted into `sys.modules['pathlib']` while at the same time importing `pathlib`.

Which I still think is a bad idea, when we have _typing we import from _typing rather than inserting it into `sys.modules['typing']`.

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: > Also which web ui is complaining about the name?

No system complains, this is 100% my own esthetical eye thing. A name like `zzz_anything` reads like "a placeholder name thrown in to test something which will be replaced with the real name later". I don't really have strong opinions about the naming, it just seemed strange to me.

Comment (mensinda) [reply mesonbuild/_pathlib.py]: The difference to `_typing` is that it is *never* actually imported in production. It is only used in `if T.TYPE_CHECKING:` blocks.

If we do the same with `pathlib` and `_pathlib` we have to replace all `import pathlib` and `from pathlib import *`, remember to always to use `_pathlib` (which can result in non-obvious errors otherwise). And once we remove the workaround we have to revert all these changes again... IMHO, at some point one dirty hack is more convenient and developer-friendly than renaming 50 different strings and enforcing that.

@jpakkane Would you then be OK, to revert to renaming it to `_pathlib` since LGTM is just broken in this regard? I really don't want it to be sorted in the middle of the other meson modules and have it blend in and look more legitimate than it is.

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: Fine by me.

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: Actually, would `import _pathlib as pathlib` work better?

Comment (mensinda) [reply mesonbuild/_pathlib.py]: If we do `import _pathlib as pathlib` or `from _pathlib import Path`, then we have to do it *everywhere* and *consitently*, which is why I wanted to do the `sys.modules['typing']` hack in the first place. This definetely is not the most elegant solution, but the soultion with the smallest git diff (not that this would make a difference after we went for fixing `encoding` on the call site...) and the least maintenance burden.

Comment (jpakkane) [reply mesonbuild/_pathlib.py]: Let's go with the current one then, it can be always fixed later. One thing to note, though, is that since this issue only occurs on Windows, the monkeypatching should probably also be done on Windows.

Comment (mensinda) [reply mesonbuild/_pathlib.py]: *If* we do `import _pathlib` we need *another* custom linter for a semi temporary fix/hack to ensure that we don't mix `_pathlib` and `pathlib`

Comment (mensinda) [reply mesonbuild/_pathlib.py:36]: Done.
